### PR TITLE
2019 May upgrade: add Schnorr opcode specification

### PR DESCRIPTION
**Pull request by markblundeberg**
This is a fleshed out specification of how to implement Schnorr, in terms of the actual opcode mechanics. A lot of discussion has been had but this is by no means settled, so more comments are welcome!

Telegram discussion group here: https://t.me/joinchat/HCYr50bcT1vAH0SkOF0Nfw

Implementation work (on Bitcoin ABC codebase) is taking place [here](https://reviews.bitcoinabc.org/T527).

### CODE REVIEW

**Comment by Mengerian** [comment]
Link to reddit discussion, for the record.

https://old.reddit.com/r/btc/comments/a7p7cl/proposal_for_may_2019_upgrade_add_schnorr/

**Comment by markblundeberg** [comment]
ABC implementation of Schnorr opcode mechanics now starting here: https://reviews.bitcoinabc.org/D2342  (until now @deadalnix has been working on the cryptographic side of this, here: https://reviews.bitcoinabc.org/D2169 )

**Comment by markblundeberg** [comment]
There has been some discussion recently about whether the flag byte is unnecessary baggage. After working on the Bitcoin ABC implementation for some time, I've found that the Schnorr opcode upgrade is straight forward enough to implement *without* the flag byte and so I've modified the spec in accordance. There is only a minor gotcha in that [implementors should take care about caching](https://github.com/bitcoincashorg/bitcoincash.org/blob/525249261a9800f3597429f0f91f4521d13907fb/spec/20190515-schnorr.md#lack-of-flag-byte----ecdsa--schnorr-ambiguity).

Some other miscellaneous changes to the spec have been made as well.

**Comment by Mengerian** [comment]
Great work Mark, Looks like the spec and implementation are coming along nicely.

I agree that it seems cleaner to not have the "flag byte" if there is no practical reason that it is needed.

**Comment by markblundeberg** [comment]
For November's upgrade, [I have an idea of how the OP_CHECKMULTISIG Schnorr support could be implemented](https://gist.github.com/markblundeberg/2c2a542d3dfbcda4d9a80157a5269d68).

During implementation, we've also found it to be clearer and more easily coded if Schnorr-sized signatures are also 'banned' from OP_CHECKMULTISIG. The spec has been updated on this small point; also, miscellaneous cleanup.

**Comment by awemany** [comment]
Hey Mark, this all sounds good though I am usually leaning conservative with the code and see the danger of having basically an 'signature is valid OR(`<ecdsa>`, `<schnorr>`)' as basically doubling the risk for the time being of a bug in either sig verification code to be disastrous. Now, I guess this is unavoidable for an upgrade of this kind.

That said and on a different note, I am wondering whether there's any way to "cross-construct" valid, if maybe nonsensical pseudo signatures that would allow a pair of `<msg> <sig>` to pass a CHECK(DATA)SIG check. Though to be honest, I am also out of my depth with respect to the various transformation that one might apply to get there (or not). From a bird's eye view, it looks to me everything should be safe and this generally impossible due to the different hash inputs in Schnorr vs. ECDSA and the impossibility for a third party to find a pre-image.

My worry is about ZCF becoming not viable any longer due to changes in the signature checking mechanisms. I think I have voiced this general and admittedly quite diffuse concern earlier.

So al I like to ask here is to please keep this - admittedly not current, though I think in the future still viable - use case and mind and like to know whether you tried to think this through (once more?) whether there's any way that such breakage might occur with that downside.

That's all. Just my 2 satoshis. Keep up the great work!

**Comment by markblundeberg** [comment]
@awemany Yes, I've kept in mind the ZCF / covenants use case the entire time while designing this system. Importatntly, those scripts only really care about two things:
* that the CHECKSIG signatures have one extra byte (hashtype) compared to the CHECKDATASIG signatures.
* that CHECKSIG internally does one more hash than CHECKDATASIG.

This ECDSA -> Schnorr replacement preserves both of those characteristics exactly. The implementation was even done in a way that should make this evident (see below). The decision to prevent 65-byte ECDSA sigs in CHECKMULTISIG even allows ZCF for multisig payments to continue unaffected, though given how messy this is, I get the feeling that you haven't developed such a ZCf multisig scheme yet :-).

(In contrast, ZCF/covenant scripts don't care about the length of the signature --- which is variable in case of ECDSA --- and neither do they care about the actual contents of the signature (the complex DER encoding), aside from the hashtype byte at the end. Thus, a drop-in replacement of the DER ECDSA signature itself with a 64 byte Schnorr signature won't affect any sensible script.)

Regarding security, yes, there's no way around the fact that every change we do creates a risk. The code changes were actually remarkably simple, however, and every change has come with a mountain of unit tests. In the ABC implementation, the crucial changes are in these:
* the Schnorr signatures themselves (in secp256k1 lib): [D2169](https://reviews.bitcoinabc.org/D2169)
* add support to Key classes in the bitcoind C++ code: [D2348](https://reviews.bitcoinabc.org/D2348)
* modify the Script interpreter's signature encoding checker to let Schnorrs through: [D2469](https://reviews.bitcoinabc.org/D2469)
* modify the Script interpreter's signature verifier to call out to Schnorr function instead of ECDSA function, where appropriate: [D2455](https://reviews.bitcoinabc.org/D2455)

There are [more refactors involved](https://reviews.bitcoinabc.org/T527) but not essential to understand what's going on.

**Comment by awemany** [comment]
Hey Mark,

thanks for the prompt and excellent answer! That certainly puts my mind at easy regarding this changeset and it makes more sense to me now that you did it like you did.

File: spec/20190515-schnorr.md [hunk 1]
```diff
@@ -1,10 +1,192 @@
...
+## Signature verification algorithm
+
+We follow [the specification from Pieter Wuille](https://github.com/sipa/bips/blob/bip-schnorr/bip-schnorr.mediawiki), but slightly modified since we allow different input pubkey encodings. Notable design choices:
+
+* Operates on secp256k1 curve.
+* Uses (R,s) Schnorr variant, not (e,s) variant.
+* Uses pubkey-prefixing in computing the internal hash.
+* The Y coordinate of R is dropped, so just its X coordinate, `r`, is serialized. The Y coordinate is uniquely reconstructed from `r` by choosing the quadratic residue.
+
+In detail, our Schnorr signature verification algorithm takes byte string `m`, non-infinite public key point `P`, and integers `r`, `s` as inputs, and does the following:
```
[the remark this answers is not in the document]
Comment (markblundeberg) [reply spec/20190515-schnorr.md]: I'm going to leave this out as it's considered common knowledge for the secp256k1 curve cryptosystem.

**Pull request merged**